doc/developer/design/20210910_include_kafka_metadata.md

```diff
@@ -0,0 +1,190 @@
+
+`INCLUDE TIMESTAMP`
+
+And an example of it being combined with existing uses of `INCLUDE KEY` full
+syntax:
+
+`INCLUDE KEY, TIMESTAMP, OFFSET, PARTITION, TOPIC`
+
+and with the renaming syntax:
+
+`INCLUDE KEY AS mykey, TIMESTAMP AS ts, OFFSET AS "Off"`
+
+### Column Types
+
+Each of the new fields will become a column in the dataflow, with an appropriate
+type:
+
+* `TIMESTAMP` will become a nullable `TIMESTAMP WITH TIME ZONE` ([javadoc][ts])
+* `OFFSET` will become a nullable `BIGINT` ([javadoc][offset])
+* `PARTITION` will become an `INTEGER` ([javadoc][partition])
+* `TOPIC` will become a `STRING` ([javadoc][topic])
+
+[ts]: https://kafka.apache.org/28/javadoc//org/apache/kafka/clients/producer/RecordMetadata.html#timestamp()
+[offset]: https://kafka.apache.org/28/javadoc//org/apache/kafka/clients/producer/RecordMetadata.html#offset()
+[partition]: https://kafka.apache.org/28/javadoc//org/apache/kafka/clients/producer/RecordMetadata.html#partition()
+[topic]: https://kafka.apache.org/28/javadoc//org/apache/kafka/clients/producer/RecordMetadata.html#topic()
+
+### Envelope Support
+
+#### Background
+
+Envelopes will have varying support for metadata fields, due variously to their
+inherent properties and engineering effort. The key constraint is that any
+envelope that knows how to do its own retractions (i.e. the Debezium and
+Materialize envelopes) require that the full retraction be present in the data.
+Essentially, retraction-aware envelopes provide `materialized` with information
+that is morally equivalent to:
+
+```javascript
+// "insert" event
+{
+    "retract": null,
+    "insert": [1, "a"]
+}
+// "update" event
+{
+    "retract": [1, "a"],
+    "insert": [1, "b"]
+}
+```
+
+The retraction-aware envelopes understand the semantics of these events and will
+correctly provide retraction/insertion messages to their dataflows.
+
+The problem comes if we add metadata, it is guaranteed that each message will be
+unique:
+
+```javascript
+// "insert" event
+{
+    "retract": null,
+    "insert": [1, "a"],
+    "offset": 100
+}
+// "update" event
+{
+    "retract": [1, "a"],
+    "insert": [1, "b"],
+    "offset": 704
+}
+```
+
+If we insert the offset from the insert event then the row in Materialize
+becomes `[1, "a", 100]`. When we see the update event, the event itself does not
+have enough information to provide a retraction, the naive retraction would be
+to try and `DELETE [1, "a", 704]`, but that does not exist and will therefore
+cause errors in dataflows.
+
+This can be worked around by maintaining UPSERT-like state for each envelope,
+but it is unclear how needed this work is, and so the next section describes the
+intended initial state and future work possible or required.
+
+#### Trivially supported
+
+`ENVELOPE NONE`: works inherently, no additional work required
+
+
+#### Small amount of work to support
+
+Both of `UPSERT` and `DEBEZIUM UPSERT` require the approximately the same amount
```

Comment:
just FYI, per Nicolle most of the customer use cases will be covered by UPSERT, so we should be aiming to ship this level of support and not just ENVELOPE NONE (but don't need to go off into the DEBEZIUM/MATERIALIZE envelopes below)